File: R/sn_univariata.R
# SN univariata (dataset, stima, diagnostica, timing)
# Primitive: dsn/psn/qsn/rsn del pacchetto 'sn'.  (API e solver in qsn: v. manuale) 


# Note: evitare start con alpha=0 (non regolare, manuale sn). 

library(sn)

set.seed(20250303)

# Parametri per la generazione
n      <- 2000
xi     <- 0
omega  <- 1
alpha  <- 1.2

# Percorsi
paths <- list(
  data = "common/data/sn_uni.csv",
  fig_dir = "common/fig",
  fig_QQ_sn = "common/fig/sn_uni_QQ_R.png",
  fig_UNIF   = "common/fig/sn_uni_UNIF_R.png",
  fig_QQ_chi2 = "common/fig/sn_uni_QQ_chi2_R.png",
  fig_profile = "common/fig/sn_uni_profile_alpha_R.png",
  tab_fit = "common/tables/sn_uni_fit_R.csv",
  log_times = "common/logs/sn_uni_times_R.csv",
  res_inversion = "common/tables/sn_uni_inversion_R.csv",
  res_unif       = "common/ables/sn_uni_unif_R.csv",
  res_reflect   = "common/tables/sn_uni_reflection_R.csv",
  res_profile   = "common/tables/sn_uni_profile_alpha_R.csv"
)

# Crea tutte le cartelle necessarie 
lapply(unique(dirname(unlist(paths))), function(d) dir.create(d, recursive = TRUE, showWarnings = FALSE))

# Funzioni di timing e formattazione
timings <- data.frame(step = character(), seconds = numeric(), stringsAsFactors = FALSE)
tic  <- function() proc.time()[["elapsed"]]
tock <- function(t0) proc.time()[["elapsed"]] - t0
add_time <- function(step, sec) {
  timings <<- rbind(timings, data.frame(step = step, seconds = sec))
}

# Log-verosimiglianza SN (DP: xi, omega, alpha)
loglik_sn <- function(par, y) {
  xi <- par[1]; om <- abs(par[2]); al <- par[3]
  sum(log(dsn(y, xi = xi, omega = om, alpha = al)))
}

# Dati: generazione + salvataggio
t0 <- tic()
y  <- rsn(n, xi = xi, omega = omega, alpha = alpha)
add_time("data_rsn_generate", tock(t0))

t0 <- tic()
write.csv(data.frame(y = y), paths$data, row.names = FALSE)
add_time("io_write_csv_data", tock(t0))

# Stima ML (trasparente)
start <- c(median(y), IQR(y)/1.349, 0.5)

t0 <- tic()
fit <- optim(
  par = start,
  fn  = function(p) -loglik_sn(p, y),
  method = "BFGS",
  control = list(reltol = 1e-10, maxit = 2000)
)
add_time("fit_ML_BFGS", tock(t0))

parhat <- c(xi = fit$par[1], omega = abs(fit$par[2]), alpha = fit$par[3])
llhat  <- -fit$value
fevals <- if (!is.null(fit$counts["function"])) as.integer(fit$counts["function"]) else NA_integer_
conv   <- fit$convergence
msg    <- if (!is.null(fit$message)) as.character(fit$message) else ""

# Diagnostica: QQ teorico SN
p <- (1:n - 0.5) / n
t0 <- tic()
q_th <- qsn(p, xi = parhat["xi"], omega = parhat["omega"], alpha = parhat["alpha"])  # solver default
add_time("qsn_vector", tock(t0))

q_emp <- sort(y)

t0 <- tic()
png(paths$fig_QQ_sn, width = 900, height = 700, res = 120)
plot(q_th, q_emp, pch = 20, main = "QQ-plot SN (R)",
     xlab = "Quantili teorici SN", ylab = "Quantili empirici")
abline(0, 1, lwd = 2, col = "gray40")
dev.off()
add_time("plot_QQ_SN_save", tock(t0))

# Diagnostica: UNIF
t0 <- tic()
u <- psn(y, xi = parhat["xi"], omega = parhat["omega"], alpha = parhat["alpha"])
add_time("psn_vector", tock(t0))

t0 <- tic()
png(paths$fig_UNIF, width = 900, height = 700, res = 120)
hist(u, breaks = 30, freq = TRUE, main = "UNIF (R)", xlab = "u = F_SN(y; θ̂)")
abline(h = length(u)/30, lwd = 2, col = "gray40")
dev.off()
add_time("plot_UNIF_save", tock(t0))

# Test KS di uniformità(0,1)
ks_p <- suppressWarnings(ks.test(u, "punif")$p.value)
write.csv(data.frame(ks_pvalue = ks_p, n = n), paths$res_unif, row.names = FALSE)

# Inversione CDF↔PPF
# Griglia spinta sulle code; usare solver="RFB" per robustezza (manuale sn).
p_grid <- c(1e-6, 1e-5, 1e-4, seq(0.001, 0.999, by = 0.001), 1 - 1e-4, 1 - 1e-5, 1 - 1e-6)

t0 <- tic()
q_inv  <- qsn(p_grid, xi = parhat["xi"], omega = parhat["omega"], alpha = parhat["alpha"], solver = "RFB")
p_back <- psn(q_inv,  xi = parhat["xi"], omega = parhat["omega"], alpha = parhat["alpha"])
add_time("invert_ppf_cdf", tock(t0))

inv_err <- data.frame(p = p_grid, p_back = p_back, abs_err = abs(p_back - p_grid))
write.csv(inv_err, paths$res_inversion, row.names = FALSE)

# QQ su z^2 vs χ^2_1
z_hat <- (y - parhat["xi"]) / parhat["omega"]
z2    <- sort(z_hat^2)
chi2_q <- qchisq((1:n - 0.5) / n, df = 1)

t0 <- tic()
png(paths$fig_QQ_chi2, width = 900, height = 700, res = 120)
plot(chi2_q, z2, pch = 20, main = "QQ di z^2 vs χ^2_1 (R)",
     xlab = "Quantili χ^2_1", ylab = "Quantili di ẑ^2")
abline(0, 1, lwd = 2, col = "gray40")
dev.off()
add_time("plot_QQ_chi2_save", tock(t0))

# Simmetria parametrica (riflessione)
y_neg <- -y
t0 <- tic()
fit_neg <- optim(
  par = c(median(y_neg), IQR(y_neg)/1.349, -0.5),
  fn  = function(p) -loglik_sn(p, y_neg),
  method = "BFGS",
  control = list(reltol = 1e-10, maxit = 2000)
)
add_time("fit_ML_reflection", tock(t0))

parhat_neg <- c(xi = fit_neg$par[1], omega = abs(fit_neg$par[2]), alpha = fit_neg$par[3])
cmp_reflect <- data.frame(
  xi_hat = parhat["xi"], xi_hat_ref = parhat_neg["xi"],
  omega_hat = parhat["omega"], omega_hat_ref = parhat_neg["omega"],
  alpha_hat = parhat["alpha"], alpha_hat_ref = parhat_neg["alpha"]
)
write.csv(cmp_reflect, paths$res_reflect, row.names = FALSE)

# Profilo di verosimiglianza in alpha
alpha_grid <- seq(-3, 3, length.out = 61)
t0 <- tic()
profile <- lapply(alpha_grid, function(a) {
  f <- function(pr) -sum(log(dsn(y, xi = pr[1], omega = abs(pr[2]), alpha = a)))
  o <- optim(c(parhat["xi"], parhat["omega"]), f, method = "BFGS",
             control = list(reltol = 1e-10, maxit = 2000))
  c(alpha = a, xi_hat = o$par[1], omega_hat = abs(o$par[2]), loglik = -o$value)
})
add_time("profile_alpha", tock(t0))

prof_df <- as.data.frame(do.call(rbind, profile))
prof_df$deviance <- 2 * (llhat - prof_df$loglik)
write.csv(prof_df, paths$res_profile, row.names = FALSE)

t0 <- tic()
png(paths$fig_profile, width = 900, height = 700, res = 120)
plot(prof_df$alpha, prof_df$deviance, type = "l", lwd = 2,
     main = "Profilo di verosimiglianza in α (R)",
     xlab = "α", ylab = "Deviance D(α)")
abline(v = 0, lty = 2, col = "gray40")
dev.off()
add_time("plot_profile_save", tock(t0))

# Salvataggi finali (stime + tempi)
out <- data.frame(
  engine = "R",
  xi_hat = parhat["xi"], omega_hat = parhat["omega"], alpha_hat = parhat["alpha"],
  loglik = llhat, n = n, fevals = fevals, convergence = conv, message = msg
)
write.csv(out, paths$tab_fit, row.names = FALSE)

write.csv(timings, paths$log_times, row.names = FALSE)

# Messaggi riassuntivi
cat("OK R — Dataset:", paths$data, "\n")
cat("OK R — Stime:",   paths$tab_fit, " (fevals =", fevals, ", conv =", conv, ")\n")
cat("OK R — Figure:",  paths$fig_QQ_sn, ",", paths$fig_UNIF, ",", paths$fig_QQ_chi2, ",", paths$fig_profile, "\n")
cat("OK R — Result tables:", paths$res_inversion, ",", paths$res_unif, ",", paths$res_reflect, ",", paths$res_profile, "\n")
cat("OK R — Tempi:",   paths$log_times, "\n")

File: Python/sn_univariata.py
# SN univariata (dataset, stima, diagnostica, timing)
# Primitive: scipy.stats.skewnorm {pdf, cdf, ppf, rvs, logpdf}

import os, time
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")  # usa un backend non interattivo (salva su file, niente GUI)
import matplotlib.pyplot as plt
from scipy import stats
from scipy.optimize import minimize

# Dizionario con tutti i percorsi usati dallo script:
#   - dataset in ingresso
#   - figure diagnostiche
#   - tabelle di risultati
#   - file di log dei tempi
paths = {
    "data":          "common/data/sn_uni.csv",
    "fig_QQ_sn":     "common/fig/sn_uni_QQ_Py.png",
    "fig_UNIF":      "common/fig/sn_uni_UNIF_Py.png",
    "fig_QQ_chi2":   "common/fig/sn_uni_QQ_chi2_Py.png",
    "fig_profile":   "common/fig/sn_uni_profile_alpha_Py.png",
    "tab_fit":       "common/tables/sn_uni_fit_Py.csv",
    "log_times":     "common/logs/sn_uni_times_Py.csv",
    "res_inversion": "common/tables/sn_uni_inversion_Py.csv",
    "res_unif":      "common/tables/sn_uni_unif_Py.csv",
    "res_reflect":   "common/tables/sn_uni_reflection_Py.csv",
    "res_profile":   "common/tables/sn_uni_profile_alpha_Py.csv",
}
# Crea tutte le cartelle necessarie (idempotente: non dà errore se già esistono)
for d in {os.path.dirname(p) for p in paths.values()}:
    if d:
        os.makedirs(d, exist_ok=True)

# lista globale che accumula i tempi per ogni step
timings = []

# funzioni tic/tock per misurare intervalli temporali
def tic() -> float: 
    return time.perf_counter()

def tock(t0: float) -> float: 
    return time.perf_counter() - t0

# aggiunge una riga (step, seconds) alla lista timings
def add_time(step: str, sec: float) -> None:
    timings.append({"step": step, "seconds": float(sec)})

# Carica dataset condiviso
# Legge il dataset SN generato dallo script R, con colonna 'y'
t0 = tic()
df = pd.read_csv(paths["data"])
add_time("io_read_csv_data", tock(t0))

# estrae il vettore y come array NumPy e la dimensione campionaria n
y = df["y"].to_numpy()
n = y.size

# ML estimation (skewnorm.fit)
# Stima dei parametri della skew-normal tramite MLE con skewnorm.fit:
#   a    ≡ alpha (slant)
#   loc  ≡ xi (location)
#   scale ≡ omega (scale)
t0 = tic()
a_hat, loc_hat, scale_hat = stats.skewnorm.fit(y)  # a ≡ alpha, loc ≡ xi, scale ≡ omega
add_time("fit_skewnorm", tock(t0))

# Calcolo della log-verosimiglianza al massimo usando la densità logaritmica
t0 = tic()
loglik = float(np.sum(stats.skewnorm.logpdf(y, a=a_hat, loc=loc_hat, scale=scale_hat)))
add_time("logpdf_sum", tock(t0))

# Diagnostica: QQ teorico SN
# Costruisce il QQ-plot dei dati rispetto alla skew-normal stimata
# 1) quantili teorici SN: q_th = F^{-1}(p; theta_hat)
# 2) quantili empirici ordinati di y
p = (np.arange(1, n + 1) - 0.5) / n
t0 = tic()
q_th = stats.skewnorm.ppf(p, a=a_hat, loc=loc_hat, scale=scale_hat)
add_time("ppf_vector", tock(t0))

q_emp = np.sort(y)
t0 = tic()
plt.figure(figsize=(7.5, 6))
plt.plot(q_th, q_emp, ".", ms=4)
plt.plot([q_th.min(), q_th.max()], [q_th.min(), q_th.max()], "-", lw=2)
plt.title("QQ-plot SN (Python)")
plt.xlabel("Quantili teorici SN")
plt.ylabel("Quantili empirici")
plt.tight_layout()
plt.savefig(paths["fig_QQ_sn"], dpi=120)
plt.close()
add_time("plot_QQ_SN_save", tock(t0))

# Diagnostica uniformità
# Calcola i PIT (Probability Integral Transform):
#   u_i = F_SN(y_i; theta_hat)
# Se il modello è ben specificato, u dovrebbe essere ~ U(0,1).
t0 = tic()
u = stats.skewnorm.cdf(y, a=a_hat, loc=loc_hat, scale=scale_hat)
add_time("cdf_vector", tock(t0))

# Istogramma di u con 30 classi
t0 = tic()
plt.figure(figsize=(7.5, 6))
plt.hist(u, bins=30)
plt.title("UNIF (Python)")
plt.xlabel("u = F_SN(y; theta_hat)")
plt.tight_layout()
plt.savefig(paths["fig_UNIF"], dpi=120)
plt.close()
add_time("plot_UNIF_save", tock(t0))

# Test KS di uniformità(0,1) sui PIT e salvataggio p-value
ks_p = float(stats.kstest(u, "uniform", args=(0, 1)).pvalue)
pd.DataFrame([{"ks_pvalue": ks_p, "n": int(n)}]).to_csv(paths["res_unif"], index=False)

# Inversione CDF↔PPF
# Verifica numerica dell'inversione CDF↔PPF:
#   1) definisce una griglia p_grid che copre core e code
#   2) calcola q_inv = PPF(p_grid)
#   3) calcola p_back = CDF(q_inv)
#   4) salva errore assoluto |p_back - p_grid|
p_grid = np.concatenate([
    [1e-6, 1e-5, 1e-4],
    np.linspace(0.001, 0.999, 999),
    [1 - 1e-4, 1 - 1e-5, 1 - 1e-6]
])
t0 = tic()
q_inv = stats.skewnorm.ppf(p_grid, a=a_hat, loc=loc_hat, scale=scale_hat)
p_back = stats.skewnorm.cdf(q_inv, a=a_hat, loc=loc_hat, scale=scale_hat)
add_time("invert_ppf_cdf", tock(t0))

inv_err = pd.DataFrame({
    "p": p_grid,
    "p_back": p_back,
    "abs_err": np.abs(p_back - p_grid)
})
inv_err.to_csv(paths["res_inversion"], index=False)

# QQ su z^2 vs chi^2_1
# Trasforma i dati tramite lo z-score stimato:
#   z_hat = (y - xi_hat) / omega_hat
# poi considera z_hat^2 e lo confronta con la distribuzione chi^2_1 via QQ-plot.
z_hat = (y - loc_hat) / scale_hat
z2 = np.sort(z_hat ** 2)
chi2_q = stats.chi2(df=1).ppf((np.arange(1, n + 1) - 0.5) / n)

t0 = tic()
plt.figure(figsize=(7.5, 6))
plt.plot(chi2_q, z2, ".", ms=4)
plt.plot([chi2_q.min(), chi2_q.max()], [chi2_q.min(), chi2_q.max()], "-", lw=2)
plt.title("QQ di z^2 vs chi^2_1 (Python)")
plt.xlabel("Quantili chi^2_1")
plt.ylabel("Quantili di z_hat^2")
plt.tight_layout()
plt.savefig(paths["fig_QQ_chi2"], dpi=120)
plt.close()
add_time("plot_QQ_chi2_save", tock(t0))

# Simmetria parametrica (riflessione)
# Riflette i dati y → -y e rifitta la skew-normal:
# ci si aspetta, idealmente:
#   xi_ref ≈ -xi_hat
#   omega_ref ≈ omega_hat
#   alpha_ref ≈ -alpha_hat
y_neg = -y
t0 = tic()
a2, loc2, scale2 = stats.skewnorm.fit(y_neg)
add_time("fit_ML_reflection", tock(t0))

# Confronto tra stime originali e stime sui dati riflessi
cmp_reflect = pd.DataFrame([{
    "xi_hat": float(loc_hat),       "xi_hat_ref": float(loc2),
    "omega_hat": float(scale_hat),  "omega_hat_ref": float(scale2),
    "alpha_hat": float(a_hat),      "alpha_hat_ref": float(a2),
}])
cmp_reflect.to_csv(paths["res_reflect"], index=False)

# Profilo di verosimiglianza in alpha:
#   - fissa alpha su una griglia
#   - per ogni alpha, ottimizza (xi, log(omega)) con BFGS
#   - calcola loglik(alpha) e la deviance rispetto al massimo globale
alpha_grid = np.linspace(-3.0, 3.0, 61)
prof = []

def nll_cond_alpha(a, par):
    """Negativa log-verosimiglianza condizionata su alpha, con parametrizzazione (xi, log(omega))."""
    xi = par[0]
    om = np.exp(par[1])  # vincolo omega>0 tramite esponenziale
    return -np.sum(stats.skewnorm.logpdf(y, a=a, loc=xi, scale=om))

t0 = tic()
for a in alpha_grid:
    # punto iniziale dell'ottimizzazione: stime globali (loc_hat, log(scale_hat))
    x0 = np.array([loc_hat, np.log(scale_hat)])
    # ottimizzazione BFGS su (xi, log(omega)) mantenendo alpha fissato
    res = minimize(
        lambda v: nll_cond_alpha(a, v),
        x0,
        method="BFGS",
        options={"maxiter": 2000, "gtol": 1e-8}
    )
    xi_hat_a = float(res.x[0])
    om_hat_a = float(np.exp(res.x[1]))
    ll_a     = -float(res.fun)
    prof.append((a, xi_hat_a, om_hat_a, ll_a))
add_time("profile_alpha", tock(t0))

# Costruisce DataFrame con il profilo: alpha, xi_hat(alpha), omega_hat(alpha), loglik(alpha)
prof_df = pd.DataFrame(prof, columns=["alpha", "xi_hat", "omega_hat", "loglik"])
# Deviance rispetto al massimo globale: D(alpha) = 2[loglik_max - loglik(alpha)]
prof_df["deviance"] = 2.0 * (loglik - prof_df["loglik"])
prof_df.to_csv(paths["res_profile"], index=False)

# Grafico del profilo di verosimiglianza in alpha
t0 = tic()
plt.figure(figsize=(7.5, 6))
plt.plot(prof_df["alpha"], prof_df["deviance"], "-")
plt.axvline(0.0, ls="--")
plt.title("Profilo di verosimiglianza in alpha (Python)")
plt.xlabel("alpha")
plt.ylabel("Deviance D(alpha)")
plt.tight_layout()
plt.savefig(paths["fig_profile"], dpi=120)
plt.close()
add_time("plot_profile_save", tock(t0))

# Salvataggi finali (stime + tempi)
# Tabella di output con stime globali e loglik
out = pd.DataFrame([{
    "engine": "Python",
    "xi_hat": float(loc_hat),
    "omega_hat": float(scale_hat),
    "alpha_hat": float(a_hat),
    "loglik": float(loglik),
    "n": int(n)
}])
out.to_csv(paths["tab_fit"], index=False)

# Salva il log dei tempi
pd.DataFrame(timings).to_csv(paths["log_times"], index=False)

# Messaggi riassuntivi a video
print("OK Python — Dataset:", paths["data"])
print("OK Python — Stime:",   paths["tab_fit"])
print("OK Python — Figure:",  paths["fig_QQ_sn"], ",", paths["fig_UNIF"], ",", paths["fig_QQ_chi2"], ",", paths["fig_profile"])
print("OK Python — Result tables:", paths["res_inversion"], ",", paths["res_unif"], ",", paths["res_reflect"], ",", paths["res_profile"])
print("OK Python — Tempi:",   paths["log_times"])
